Step2

gaeul-kim opened the pull request:
- Car name 속성 추가
- UserRecord 추가
- ConsoleMain display 로직 분리
- 우승자 선정 로직 추가

게임기록의 자료구조를 고민했는데 만족할만한 답을 찾지 못했습니다.
그래서 우승자 선정하는 로직도 지나치게 복잡해지고 테스트 케이스 작성이 잘 안된것 같습니다.
피드백 받고 좀더 리팩토링 해보겠습니다.

gaeul-kim committed:
step2 first commit

- Car name 속성 추가
- UserRecord 추가
- ConsoleMain display 로직 분리

gaeul-kim committed:
우승자 선정 로직추가

javajigi reviewed (changes requested):
도메인 로직과 View 로직을 구분하기 위해 한번 시도할 때마다의 출력 결과를 모두 기록하고 있는데요.
출력 결과를 모두 기록하지 말고 다음과 같이 구현하는 것은 어떨까요?

```
public static void main(String[] args) {
    int carNo = InputView.getCarNo();
    int tryNo = InputView.getTryNo();

    RacingGame racingGame = new RacingGame(carNo);
    GameResult result;
    for(int i = 0; i < tryNo; i++) {
        result = racingGame.move();
        ResultView.print(result);
    }
    ResultView.printWinner(result);
}
```

이렇게 구현하면 GameResult는 List<Car> 정보만 가지고 있으면 되고, 우승자에 관련한 로직은 GameResult에서 List<Car> 정보를 활용해 구현하면 어떨까요?

gaeul-kim committed:
게임 진행방식 변경

- RacingGame 내부가 아닌 Main 클래스에서 반복하도록 변경
- 회차별 기록을 저장하지 않도록 변경
- View로직(Input, Print) 클래스 분리

피드백을 참고하여 게임 구조를 전체적으로 변경하였습니다.
객체별 역할이 좀더 명확하게 구분되는 효과가 있는것 같습니다.
테스트 케이스는 좀더 고민을 해보겠습니다.

javajigi commented on src/main/java/racingcar/GameResult.java:
```diff
@@ -0,0 +1,41 @@
+                winners.add(record);
+            }
+        }
+        return winners;
```
메소드가 너무 많은 일을 하고 있다. 다음 원칙을 지켜 메소드를 구현한다.

작게 만들어라. - 함수를 만드는 첫 번재 규칙은 '작게'다. 함수를 만드는 두 번째 규칙은 '더 작게'다.
한 가지만 해라. - 함수는 한 가지를 해야 한다. 그 한 가지를 잘 해야 한다. 그 한 가지만 해야 한다.

javajigi commented on src/main/java/racingcar/GameResult.java:
```diff
@@ -0,0 +1,41 @@
+        Record maxRecord = this.records.stream().max(Record::compareTo).get();
+        List<Record> winners = new ArrayList<>();
+        for (Record record : records) {
+            if (record.getPosition() == maxRecord.getPosition()) {
```
상태를 가지는 객체의 데이터를 꺼내려(get)하지 말고 객체에 메시지를 보내라는 원칙에 따라 구현한다.

지금 구조와 maxRecord.matchPosition(record)와 같이 구현하는 것 중 어느 것이 나을까?

javajigi commented on src/main/java/racingcar/interfaces/Car.java:
```diff
@@ -1,14 +1,14 @@
-    Integer getCurrentPosition();
+    Integer getPosition();
 
     String getName();
```
Record에 position과 name을 담고 있기 때문에 move 메소드만 있어도 되지 않을까?

hello-sangsik replied:
Record의 생성자에 Car 인스턴스를 사용하여 값을 넣어주는데 메소드가 불필요할까요?
아니면 인터페이스에서 삭제하고 구현클래스에서만 선언하라는 뜻인지 궁금합니다.

javajigi reviewed (changes requested):
구현 깔끔하게 잘 했네요. step2 마지막 피드백이 될 것 같아요.

gaeul-kim committed:
step2 refactoring

- Car : getName, getPosition 메소드 삭제
- GameResult : 우승자 결정 로직에 스트림 적용
- Record : 상태를 비교하는 방법 수정
- 기타 코드, 변수명 정리
